### DISCUSSION TITLE
Sublime PML Needs Your Feedback!

### DISCUSSION THREAD

**Discussion by tajmone** [tajmone/Sublime-PML#45] [General]
Since the recent bump to PML/PMLC v4.0.0 I've been working to ensure that **Sublime PML** works with the latest PML version.

I'm also glad to announce that **Sublime PML** is finally starting to gain pace toward maturity. Although there are still tons of unsupported PML nodes (which fallback on generic "unknown node" scoping), the design of the package is proving its robustness and those nodes which are implemented are greatly benefiting from the "smart scoping" features on which I've invested hard since inception.

The reason development has been so slow is due to the _deliberate_ choice on focusing on strong semantics from the onset. The goal has always been to carefully provide extra meta-scopes in order to gain fine-grain control over completions and snippets suggestions, to prevent "context pollution" — i.e. non-pertinent suggestions showing up in the wrong places.

While on the end-user side this might seem obvious an simple, the opposite is true from the back-end/developer perspective. Obtaining fine-grain control of semantic contexts via "dumb" single-line RegExs is all but simple, especially with the PML syntax, where everything is either a node, attributes or contents.

I'm fairly confident that the current state of **Sublime PML** offers a good end-user experience — albeit incomplete, at least the supported feature are well implemented. Being a lightweight markup daily user myself, I'm well aware of how a syntax package can either render your editing experience a breeze or hell; so I've striven hard to ensure that **Sublime PML** will end up being a good package, friendly to PML editors.

But at this point, **Sublime PML** really needs your feedback in order for me to ensure that this enthusiasm is shared by the rest of its users, and to obtain valuable feedback on how to improve it.

Start by reading the _[Sublime PML User Guide]_ — which can be quickly accessed from within Sublime Text via the menu:

- **Tools** &gt; **PML** &gt; **Sublime PML User Guide**

The guide should provide you useful insights into the keyboard shortcuts and snippets/completions triggers that will improve your editing experiences.

All the shortcuts and triggers are designed to be context-specific, so if you see an out-of-place suggestion (e.g. a formatting key-shortcut or completion/snippet suggestion working in a raw-node or within an `[options` block) please report it as a bug.

Keeping completions and key-shortcuts context specifics means that ultimately it will be possible to add tons of these features to the package without fearing context pollution. If you've used a poor quality package before, where dozens of completions simply show up all over the place while you type, you'll know what I mean by the term "context pollution" — i.e. that these features cease to be useful since the desired functionality is simply lost in the resulting clutter.

After much experimenting, rolling back and trying better patterns, I'm now fairly confident that I've nailed the proper approach to handling PML syntax scoping, and I can start to focus on adding the rest of the PML nodes which still lack specific support.

Some advanced features, like footnotes and cross-references, might require some thorough planning before they are added to the package, because I want them to be well supported, so that editors can work with them in a seamless manner.

There are some other PML nodes for which I've postponed their implementation simply because I'm aware (from my exchanges with @pml-lang) that they will be undergoing breaking changes in the nearby future — so I'm trying to avoid having to amend or rewrite the syntax as much as possible.

Overall, I think that **Sublime PML** already offers a reasonable editing experience, and I've been using it in production since its creation, which has allowed to me to prioritize addition of missing features based on direct usage experience.

But now I'd like to hear feedback from the **Sublime PML** userbase — to be honest, I don't even know if there's anyone beside me using this package on a daily base to edit PML documents. So this would be a good occasion for users of this package to introduce themselves, and start forming a community around the package.

Tristano.

[Sublime PML User Guide]: https://tajmone.github.io/Sublime-PML "Read the Sublime PML User Guide online"

**Comment by pml-lang** [comment]
This is excellent news !!!

You are indeed investing time and energy to get it right and create the _best possible_ user experience. Surely this is very much appreciated by all kinds of users. Moreover, your plugin can later be used as a starting point for other plugins and/or other dialects of PDML. 👌

> I've postponed their implementation simply because I'm aware (from my exchanges with @pml-lang) that they will be undergoing breaking changes in the nearby future

There are currently no _planned_ breaking changes, because they have all been implemented in PMLC version 4.0.0.
However, _other_ breaking changes (not planeed yet) will likely occur in the future.

> now I'd like to hear feedback from the Sublime PML userbase

Yes. As said already, PML promotion will start this year, so users should soon send feedback. I'm going to use Sublime-PML too and send feedback.

Note: The 'available on GitHub' link in the [Sublime PML User Guide](https://tajmone.github.io/Sublime-PML/) still points to the old repository name (https://github.com/pml-lang/converter) (but it's automatically redirected to the new name). I suggest to use the new name in the link:
https://github.com/pml-lang/pml-companion.

**Reply by tajmone** [reply]
> Note: The 'available on GitHub' link in the Sublime PML User Guide still points to the old repository name (https://github.com/pml-lang/converter) (but it's automatically redirected to the new name). I suggest to use the new name in the link:
> https://github.com/pml-lang/pml-companion.

Thanks for pointing it out @pml-lang! I've now fixed all links pointing to old repo names (caught some others too), both in READMEs and the User Guide.


> There are currently no planned breaking changes, because they have all been implemented in PMLC version 4.0.0.
> However, other breaking changes (not planeed yet) will likely occur in the future.

It's the latter I was referring to, i.e. those features which we know that aren't yet fully fleshed out (or, indeed, are provisional) and are very likely to undergo changes at some point.

> I'm going to use Sublime-PML too and send feedback.

That would be much appreciated, since coming up with novel ways to simplify editing requires multiple opinions and feedback on how users go about the tasks, which would enable to simplify the required steps.
